[Zechary Au] iP

zechajw opened the pull request:
### Duke

> "A goal without a plan is just a wish" - Antonie de Saint-Exupéry

Duke is an _easy to use_, **CLI based** calendar and task planner. It's

- easy to learn and use
- has a compact file size
- downloadable for offline use

It's features include:
- [X] Adding Tasks to a List
- [X] Categorizing Tasks as Todos, Events or Deadlines
- [X]  Finding Tasks by keyword
- [X] Saving Tasks to an offline file
- [X] Cross-machine compatibility using simple text save files.

All you need to do is

1. Click the link to download the file
2. Open the app to start using **TODAY**

Simply open the application and type `todo Hello World` to try it out [today](https://github.com/nus-cs2103-AY2122S2/ip/pull/18)!

Plan your life with ~~difficulty~~ absolute ease today! :hugs:

This code was written completely in Java
Sample Code:
```java
class Duke {
    public static void main(String[] args) {
        System.out.println("Hello World!");
    }
}
```

Commit:
Add Gradle support

Commit:
Add .class files to gitignore, add greet echo exit

Commit:
Add Calendar, Task for Add, List functionality

Commit:
Add mark, unmark functionality for Level 3

To ensure that there is no invalid input, command checks for if the
number is within range of list and if the value input after
"mark/unmark" is an integer value

Commit:
Split tasks into Deadline, Event, Todo for Level-4

Commit:
Added Exception Handling and readded mark/unmark

Commit:
Add delete function with exception handling

Commit:
Automated Text UI Testing

Commit:
Organize previous code into packages

Previously, the bulk of the code resided in the main Duke driver class,
and OOP was only employed to model the Calendar and Tasks (Deadline,
Todo, Event).

To better to adhere to OOP principles for this software engineering
project. The code was refactored into different packages. All the code
is in the overarching duke package, and each subpackage aims to serve a
specific function for the program. E.g. duke.info stores all the
information required for the program (Calendar, Tasks, Exceptions) and
the duke.ui package handles the reading and writing from user input.

Let's
* Refactor the code into separate packages for better organization
* Use parent abstract classes to abstract out common functionalities
* Use a Text file to store all standard output messages

zechajw committed:
Merge pull request #1 from zechajw/branch-Level-7

Organize previous code into packages

Commit:
Fix bug in "unmark" command

Previously, there was a bug where the unmark command did not mark the
job as incomplete, but instead marked the job as complete due to an
erroneous call of markAsDone() instead of markAsNotDone().

This commit changes the method call to properly reflect the action of
marking the task as incomplete.

Let's
* Fix the unmark command to properly carry out the desired functionality

zechajw committed:
Merge pull request #2 from zechajw/branch-Level-7

Fix bug in "unmark" command

Commit:
Add save functionality when Calendar chanes

Previously, the list of tasks in the Calendar was deleted when the
program exits. This meant that the user would lose all the information
about the tasks after each run of the program.

To make the usage of the program more meaningful, a save functionality
was added that saves the current list of tasks to "tasks.txt" in the
main program folder whenever the task list is updated. This makes the
program more meaningful to use, as there is persistent information that
users can use between each program run.

Let's
* Add a save function to allow users to keep their progress
* Add a load function to automatically load a previous save on startup
* Use a "tasks.txt" file in the main directory to store progress

zechajw committed:
Merge pull request #3 from zechajw/branch-Level-7

Add save functionality on Calendar change

Commit:
Add Date functionality to Duke

Currently, dates are stored merely as strings in each Task. This means
that the dates cannot be manipulated in a meaningful way and prevents
any further improvement to the date functionality (such as filtering
tasks by date)

To make the date formatting more useful, the date attribute now utilizes
the LocalDate library to parse dates. When a date is input, the Task()
constructor will attempt to parse it into a LocalDate format. If the
date cannot be parsed, it will be treated as a simple string.

Lets
* Use the LocalDate library to allow dates to be utilized more
  functionally

zechajw committed:
Merge pull request #4 from zechajw/branch-Level-8

Add Date functionality to Duke

Commit:
Add Find Functionality to Duke

Currently, the list of tasks can only be shown statically, which means
that as the list of tasks scale up and there are a large amount of
tasks, it would be difficult for a user to search for a specific task or
any tasks matching a certain criteria.

To make the calendar more useful, FindCommand is added and the user can
input "find <keyword>" to list all tasks that match that particular
keyword.

Lets
* Add a find function to allow users to find tasks matching a keyword

Commit:
Add JavaDocs documentation to code

Currently, there is no documentation for the codebase. This makes it
difficult for other developers to contribute to the project or to fix
bugs related to the codebase in the future.

To make the codebase more informative, JavaDocs documentation has been
added to describe each method and ease future development of the
codebase

Lets
* Add JavaDocs documentation to all methods in the codebase

zechajw committed:
Merge pull request #5 from zechajw/branch-A-JavaDoc

Branch a java doc

Commit:
Add JUnit tests for Todo

Currently, there are no unit tests for the classes in the package. This
means that there is no confirmation that any of the classes is working
as intended, which might result in regression bugs in the future when
the code is extended.

To test the codebase, JUnit tests for Todo was added to check that the
methods in the Todo class are working as intended.

Lets
* Add preliminary JUNit tests for the Todo class

zechajw committed:
Merge pull request #6 from zechajw/branch-A-JUnit

Add JUnit tests for Todo

NicsunXnus commented on src/main/java/duke/duke/Duke.java:
```diff
@@ -0,0 +1,68 @@
+
+    /**
+     * Runs the Duke program using the run() method specified.
+     * @param args - command line arguments
```
Should there be no spacing between the description and parameter section? I noticed the issue in other places also.
```suggestion
     * Runs the Duke program using the run() method specified.
     *
     * @param args - command line arguments
```

NicsunXnus commented on src/main/java/duke/duke/Duke.java:
```diff
@@ -0,0 +1,68 @@
+     * Executes the Duke program.
+     */
+
+    public void run() {
```
Should there be a spacing between the javadoc comment and the method run()?

NicsunXnus commented on src/main/java/duke/duke/commands/Command.java:
```diff
@@ -0,0 +1,35 @@
+    /**
+     * Constructs an empty Command
+     */
+    Command() {}
```
Should it all be on a single line?
```suggestion
    Command() {
    }
```

NicsunXnus commented on src/main/java/duke/duke/Duke.java:
```diff
@@ -0,0 +1,68 @@
+import duke.storage.Storage;
+
+import duke.ui.Ui;
+
```
Should there be no header comments for classes? I also noticed the same issue in several other classes.

NicsunXnus commented on src/main/java/duke/duke/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,49 @@
+            ui.showTaskDeleted(taskString);
+            storage.save(calendar);
+        } catch (IndexOutOfBoundsException e) {
+            throw new InvalidInputException(String.format(Text.TEXT_INVALID_LIST_INDEX, calendar.numOfEntries(), Integer.parseInt(indexToDelete)));
```
I think that this could be used to increase readability.
```suggestion
            throw new InvalidInputException(String.format(Text.TEXT_INVALID_LIST_INDEX
                , calendar.numOfEntries(), Integer.parseInt(indexToDelete)));
```

NicsunXnus commented on src/main/java/duke/duke/info/task/Task.java:
```diff
@@ -0,0 +1,134 @@
+    /**
+     * Marks the task as complete by setting isComplete to true
+     */
+    void complete() {
```
Should this better rephrased?
```suggestion
    void markTaskComplete() {
```

NicsunXnus commented on src/main/java/duke/duke/info/task/Task.java:
```diff
@@ -0,0 +1,134 @@
+    /**
+     * Marks the task as incomplete by setting isComplete to false
+     */
+    void incomplete() {
```
Should this be better rephrased?
```suggestion
    void markTaskIncomplete() {
```

NicsunXnus commented on src/main/java/duke/duke/ui/Ui.java:
```diff
@@ -0,0 +1,116 @@
+     * @param addedTask - the task that was added
+     * @param numOfTasks - the total number of tasks
+     */
+    public void showTaskAdded(Task addedTask, int numOfTasks) { System.out.println(String.format(Text.TEXT_TASK_ADDED, addedTask, numOfTasks)); }
```
Should this be all on a single line? I also noticed the same issue for some methods in this class.

NicsunXnus commented on src/test/java/TodoTest.java:
```diff
@@ -0,0 +1,19 @@
+    }
+
+    @Test
+    public void StringTest() {
```
Should this be better rephrased, as String is generally kind of a vague description?
```suggestion
    public void toStringTest() {
```

NicsunXnus reviewed (commented):
Almost there! Some changes required such as adding header comments to the Classes, and some other minor ones.

Halpfrog commented on src/main/java/duke/duke/commands/AddCommand.java:
```diff
@@ -11,10 +11,23 @@
     private final Task taskToAdd;
 
+    /**
+     * Constructs an AddCommand with the specified taskToAdd, which contains
+     * the task to be added in the execute command.
+     * @param taskToAdd - the task to be added
```
Should there be a gap between the description and param?

Halpfrog replied:
perhaps leave a gap after the description

Halpfrog commented on src/main/java/duke/duke/commands/AddCommand.java:
```diff
@@ -11,10 +11,23 @@
     }
 
+    /**
+     * Adds the task specified in taskToAdd to the specified calendar. Afterwards,
+     * display a success message showing that the task has been successfully added
+     * using the ui handler, and save the new version of the calendar to storage.
+     * @param calendar - the calendar used in the program
```
Should there be a gap after the description?

Halpfrog commented:
Great work overall! Just spotted some issues with spacing but other than that well done

Commit:
Manifest

Commit:
Merge remote-tracking branch 'origin/add-gradle-support' into Week-4

Commit:
Add Checkstyle and gradle support

Commit:
Added Duke UI

zechajw committed:
Merge pull request #7 from zechajw/Week-4

Week 4

zechajw committed:
Merge pull request #8 from zechajw/javafx

Added Duke UI

Commit:
Add Java Assertions

Currently, there are not assertions for the methods in the package. This
means that the integrity of the code is not checked as the code runs,
and there could be unintended bugs at various points in the code. This
might result in regression bugs in the future when the code is extended.

To test the codebase, Java assertions were added at various points to
check that the methods are working as intended.

Lets
* Add assertions to ensure that the code is working at various points

zechajw committed:
Merge pull request #10 from zechajw/branch-assertions

Add Java Assertions

Commit:
Added Recurring Events functionality

Currently, there is no support for recurring event functionality in
Duke. However, users might want to schedule an event that recurs at a
specific interval (interval of 7 days). The lack of this feature would
mean that users would have to manually calculate and input the events
one-by-one until the desired effect is achieved.

Using the LocalDate feature, users can now use the "recurring event"
functionality to add a recurring event to their Calendar. The command
takes the form "recurring event <action> /at <date> /interval <number>
/ends <end date>. Each parameter is delimited by the appropriate /label
to separate each input for easier parsing.

Lets
* Add a preliminary recurring event functionality

zechajw committed:
Merge pull request #11 from zechajw/branch-BCD-Extension

Added Recurring Events functionality

Commit:
Added level-10 branch from previous week

zechajw committed:
Merge pull request #12 from zechajw/branch-Level-10

Added level-10 branch from previous week

Commit:
Add Ui.png to docs

zechajw committed:
Merge pull request #13 from zechajw/branch-A-CodeQuality

Add Ui.png to docs

zechajw committed:
Set theme jekyll-theme-slate

Commit:
Update README.md in docs folder for github pages

Commit:
Merge branch 'master' of https://github.com/zechajw/ip

Commit:
Fix checkstyle error in Task.java

zechajw committed:
Set theme jekyll-theme-cayman

zechajw committed:
Set theme jekyll-theme-cayman
